### PR TITLE
cpu/sam0_common: move PWM to common code, add support for saml21, samd5x

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

This moves the `pwm.c` driver from `samd21` to `sam0_common` and adds the required clock setup for the newer sam0 members.

While touching the PWM configuration anyway I also made the GCLK source configurable and added convenience defines to set the timer configuration.

Unfortunately I wasn't able to get the PWM to work with TC (instead of TCC), so no PWM for saml1x yet.

### Testing procedure

Flash `tests/periph_pwm`, the `osci` command should flash the on-board LED.

Only tested on `same54-xpro`.

### CODE REVIEW

File: boards/same54-xpro/include/periph_conf.h [hunk 1]
```diff
@@ -148,6 +148,35 @@ static const uart_conf_t uart_config[] = {
...
+ * @name PWM configuration
+ * @{
+ */
+#define PWM_0_EN            1
+
+#if PWM_0_EN
+/* PWM0 channels */
+static const pwm_conf_chan_t pwm_chan0_config[] = {
+    /* GPIO pin, MUX value, TCC channel */
+    { GPIO_PIN(PC, 18), GPIO_MUX_F, 2 },
```
Comment (dylad) [inline boards/same54-xpro/include/periph_conf.h:161]: Why choosing this pin ? It is not available on expander connectors.
PC18 is only available through DGI or PDEC on SAME54.

Comment (benpicco) [reply boards/same54-xpro/include/periph_conf.h:161]: It's the on-board LED - so you can get immediate feedback with the `osci` command.

**Comment by dylad** [comment]
Something is wrong with SAML21. I connected an external LED to the PWM pin but it doesn't oscillate. LED is just on and that's it.

**Comment by benpicco** [comment]
Huh, then something is wrong there. Can you try with `SAM0_GCLK_MAIN` instead of `SAM0_GCLK_8MHZ`?

Btw PB10 should be the on-board LED as well, so no need to connect anything.

File: cpu/sam0_common/include/timer_config.h [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+#define MCLK_TCC3_MASK (MCLK_APBDMASK_TCC3)
+#endif
+/** @} */
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* TIMER_CONFIG_H */
+/** @} */
```
Comment (dylad) [inline cpu/sam0_common/include/timer_config.h:221]: What's the point of this file ?

File: cpu/sam0_common/periph/pwm.c [hunk 3]
```diff
@@ -126,11 +73,33 @@ static uint8_t get_prescaler(unsigned int target, int *scale)
...
+static void poweroff(pwm_t dev)
+{
+    const pwm_conf_t *cfg = &pwm_config[dev];
+
+#ifdef MCLK
+    GCLK->PCHCTRL[cfg->tim.gclk_id].reg = 0;
+    *cfg->tim.mclk &= ~cfg->tim.mclk_mask;
+#else
+    PM->APBCMASK.reg &= ~cfg->tim.pm_mask;
+    GCLK->CLKCTRL.reg = GCLK_CLKCTRL_GEN_GCLK7
```
Comment (dylad) [inline cpu/sam0_common/periph/pwm.c:100]: Why GEN_GCLK7 ?

File: cpu/sam0_common/include/timer_config.h (hunk 2 above)
Comment (benpicco) [reply cpu/sam0_common/include/timer_config.h:221]: To allow for a generic `TCC_CONFIG` (and later `TC_CONFIG`) macro.

**Comment by dylad** [comment]
> Can you try with SAM0_GCLK_MAIN instead of SAM0_GCLK_8MHZ?

Doesn't change anything.

File: cpu/sam0_common/periph/pwm.c (hunk 3 above)
Comment (benpicco) [reply cpu/sam0_common/periph/pwm.c:100]: I guess the idea was to connect it to an unused GCLK so the original one can be turned off if it runs ondemand.
This was there already, I just kept that part - I can check if it makes any difference to power consumption (if something other than `SAM0_GCLK_MAIN` is used), I'd be happy to drop it.

File: cpu/sam0_common/periph/pwm.c [hunk 4]
```diff
@@ -17,79 +17,26 @@
...
-    if (tcc == TCC3) {
-        return PM_APBCMASK_TCC3;
-    }
-#endif
-
-    assert(0);
-    return 0;
-}
-
 static uint8_t get_prescaler(unsigned int target, int *scale)
```
Comment (dylad) [inline cpu/sam0_common/periph/pwm.c:40]:
```suggestion
static uint8_t _get_prescaler(unsigned int target, int *scale)
```

**Comment by benpicco** [comment]
> Could you also rename all the _tcc internal reference (since we can use TC) please ?

What do you mean by that? TCC and TC can't be used interchangeably, they don't have an identical register set or features.
Both can be used for PWM generation, but configuration will differ.

**Comment by benpicco** [comment]
I can just drop the `saml21-xpro` config for now.

**Comment by benpicco** [comment]
I removed the patch for `saml21-xpro`.

<details><summary>saml21_xpro.patch</summary>

```patch
diff --git a/boards/saml21-xpro/Makefile.features b/boards/saml21-xpro/Makefile.features
index 12d5fdbcfa..b397b59555 100644
--- a/boards/saml21-xpro/Makefile.features
+++ b/boards/saml21-xpro/Makefile.features
@@ -5,6 +5,7 @@ CPU_MODEL = saml21j18a
 FEATURES_PROVIDED += periph_adc
 FEATURES_PROVIDED += periph_dac
 FEATURES_PROVIDED += periph_i2c
+FEATURES_PROVIDED += periph_pwm
 FEATURES_PROVIDED += periph_rtc
 FEATURES_PROVIDED += periph_rtt
 FEATURES_PROVIDED += periph_spi
diff --git a/boards/saml21-xpro/include/periph_conf.h b/boards/saml21-xpro/include/periph_conf.h
index c66a7d4898..c4983db1c6 100644
--- a/boards/saml21-xpro/include/periph_conf.h
+++ b/boards/saml21-xpro/include/periph_conf.h
@@ -105,6 +105,35 @@ static const uart_conf_t uart_config[] = {
 #define UART_NUMOF          ARRAY_SIZE(uart_config)
 /** @} */
 
+/**
+ * @name PWM configuration
+ * @{
+ */
+#define PWM_0_EN            1
+
+#if PWM_0_EN
+/* PWM0 channels */
+static const pwm_conf_chan_t pwm_chan0_config[] = {
+    /* GPIO pin, MUX value, TCC channel */
+    { GPIO_PIN(PB, 10), GPIO_MUX_F, 4 },
+};
+#endif
+
+/* PWM device configuration */
+static const pwm_conf_t pwm_config[] = {
+#if PWM_0_EN
+    { .tim  = TCC_CONFIG(TCC0),
+      .chan = pwm_chan0_config,
+      .chan_numof = ARRAY_SIZE(pwm_chan0_config),
+      .gclk_src = SAM0_GCLK_8MHZ,
+    },
+#endif
+};
+
+/* number of devices that are actually defined */
+#define PWM_NUMOF           ARRAY_SIZE(pwm_config)
+/** @} */
+
 /**
  * @name    SPI configuration
  * @{
```
</details>

**Comment by benpicco** [comment]
I figured out what the problem was - saml21 only has 4 CC channels, so the output channels will wrap.
They can be re-mapped using the OTMX register, but I think this is not needed - the LED on the `saml21-xpro` works now.

File: cpu/sam0_common/include/periph_cpu_common.h [hunk 5]
```diff
@@ -212,6 +213,57 @@ typedef struct {
...
+    gpio_mux_t mux;         /**< pin function multiplex value */
+    uint8_t chan;           /**< TCC channel to use */
+} pwm_conf_chan_t;
+
+/**
+ * @brief   PWM device configuration data structure
+ */
+typedef struct {
+    tcc_cfg_t tim;              /**< timer configuration */
+    const pwm_conf_chan_t *chan;/**< channel configuration */
```
Comment (dylad) [inline cpu/sam0_common/include/periph_cpu_common.h:262]:
```suggestion
    const pwm_conf_chan_t *chan;    /**< channel configuration */
```

**Comment by dylad** [comment]
I can give it a try on SAML21,SAMD54 and maybe SAMD21 tomorrow.

File: cpu/sam0_common/periph/pwm.c [hunk 6]
```diff
@@ -17,80 +17,42 @@
...
+#elif defined(TCC_SYNCBUSY_CC3)
+#define CC_NUMOF    4
+#elif defined(TCC_SYNCBUSY_CC2)
+#define CC_NUMOF    3
+#elif defined(TCC_SYNCBUSY_CC1)
+#define CC_NUMOF    2
+#elif defined(TCC_SYNCBUSY_CC0)
+#define CC_NUMOF    1
+#endif
+
```
Comment (dylad) [inline cpu/sam0_common/periph/pwm.c:44]: Can't we use TCCx_CC_NUM to determine the number of CC instead ?

Comment (benpicco) [reply cpu/sam0_common/periph/pwm.c:44]: You mean instead of `CC_NUMOF`?

Comment (benpicco) [reply cpu/sam0_common/periph/pwm.c:44]: Ah I see what you mean

Comment (benpicco) [reply cpu/sam0_common/periph/pwm.c]: Thank you, that uncovered a bug - I miss-understood CCB and the amout of Capture/Compare channels is not uniform across timers.

**Comment by dylad** [comment]
This still works, please squash !

**Comment by benpicco** [comment]
Squashed.

**Pull request merged**